staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+			test.handlerErrVerifier.verify(t, err)
+		})
+	}
+}
+
+func TestPerRequestWriteDeadlineWithTimeoutOccursBeforeHandlerWrites(t *testing.T) {
+	// This test documents the behavior of the per request write deadline
+	// using a standard net http server.
+	//
+	// scenario: the handler writes to and flushes the ResponseWriter object
+	// after write timeout occurs:
+	//  - a) write deadline is set to 100ms
+	//  - b) the handler waits until the write deadline exceeds
+	//  - c) the handler writes a few bytes so there is some content to be flushed
+	//  - d) the handler invokes FlushError
+	// timeout occurs before the handler writes to the ResponseWriter object,
+	// so we expect the following:
+	//    e) client: expected to receive an error from the server
+	//    f) server: FlushError method of the ResponseWriter object should
+	//       return an "i/o timeout" error immediately.
+	t.Parallel()
+
+	const deadline = 100 * time.Millisecond
+	tests := []struct {
+		protoMajor         int
+		waiter             waiter
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1,
+			// write timeout is set to 100ms, a wait of 5s should
+			// be enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+		{
+			protoMajor:         2,
+			waiter:             &waitWithChannelClose{after: make(chan time.Time)},
+			clientErrVerifier:  wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			handlerErrCh := make(chanErr, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerErrCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("server: expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
+					return
+				}
+				flusher, ok := w.(interface{ FlushError() error })
+				if !ok {
+					t.Errorf("server: expected ResponseWriter object to implement FlushError")
+					return
+				}
+
+				// a) set write deadline for this request
+				ctrl := http.NewResponseController(w)
+				if err := ctrl.SetWriteDeadline(time.Now().Add(deadline)); err != nil {
+					t.Errorf("server: unexpected error from SetWriteDeadline: %v", err)
+					return
+				}
+
+				// b) wait until the write deadline exceeds
+				<-test.waiter.wait()
+
+				// c) write a few bytes so there is something in the buffer to
+				// flush, the client should never see these bytes in the response.
+				if _, err := w.Write([]byte("hello")); err != nil {
+					t.Errorf("server: unexpected error from Write after timeout: %v", err)
```

Comment:
why is this guaranteed to succeed and the 1kb writes above are not? Is Golang specifying some buffer size?